### ISSUE TITLE
npm run test

### ISSUE THREAD

**Issue by mig-25** [garris/BackstopJS#240]
I am sitting with macOS and with  node v4.4.7 and npm  v3.10.5
I have followed the example on css-tricks. When running the firs test without any changes, its generating the files accordingly, but it launches my Windows virtual machine to launch http://localhost:3001/compare/ but as a broken page . If I run the same url in my mac, it is also is broken. 
But I can see in the terminal that everything passes. After making the suggested changes to the css code, and I run npm run test, once again it launches the browser but the link is broken.

I can see in the terminal that the test failed. But after the Finished 'openReport' after 12 ms lot of errors are reported as included in this  message.

```
MacBook-Pro:backstopjs sohail$ npm run test

> backstopjs@1.3.5 test /Users/sohail/GitHub/test/node_modules/backstopjs
> gulp test


BackstopJS Config loaded at location /Users/sohail/GitHub/test/backstop.json
[17:23:08] Using gulpfile ~/GitHub/test/node_modules/backstopjs/gulpfile.js
[17:23:08] Starting 'init'...
[17:23:08] Finished 'init' after 71 μs
[17:23:08] Starting 'test'...

Running CasperJS with:  [ 'capture/genBitmaps.js' ]
[17:23:08] Finished 'test' after 8.32 ms
CasperJS: Ready event received.
CasperJS: Current location is ../../index.html
CasperJS: Screenshots for phone (320x480)
CasperJS: Ready event received.
CasperJS: Current location is ../../index.html
CasperJS: Screenshots for tablet_v (568x1024)
CasperJS: Ready event received.
CasperJS: Current location is ../../index.html
CasperJS: Screenshots for tablet_h (1024x768)
CasperJS: Comparison config file updated.

Bitmap file generation completed.
gulp.run() has been deprecated. Use task dependencies or gulp.watch task triggering instead.
[17:23:14] Starting 'start'...

Server launched in background with PID: 43424
Listening on port: 3001
NOTE: Sever will auto-shutdown in 15 mins.

[17:23:15] Finished 'start' after 1.01 s
[17:23:15] Starting 'report'...
[17:23:15] Finished 'report' after 3.65 ms
[17:23:16] Starting 'compare'...
 OK: My Local Test 0_5_footer_0_phone.png 
 OK: My Local Test 0_5_footer_1_tablet_v.png 
 OK: My Local Test 0_0_nav_0_phone.png 
 OK: My Local Test 0_5_footer_2_tablet_h.png 
 OK: My Local Test 0_0_nav_1_tablet_v.png 
 OK: My Local Test 0_0_nav_2_tablet_h.png 
 OK: My Local Test 0_2_bodycol-md-4nth-of-type1_0_phone.png 
 OK: My Local Test 0_3_bodycol-md-4nth-of-type2_0_phone.png 
 OK: My Local Test 0_4_bodycol-md-4nth-of-type3_0_phone.png 
 OK: My Local Test 0_2_bodycol-md-4nth-of-type1_2_tablet_h.png 
 OK: My Local Test 0_4_bodycol-md-4nth-of-type3_2_tablet_h.png 
 OK: My Local Test 0_3_bodycol-md-4nth-of-type2_2_tablet_h.png 
 OK: My Local Test 0_2_bodycol-md-4nth-of-type1_1_tablet_v.png 
 OK: My Local Test 0_3_bodycol-md-4nth-of-type2_1_tablet_v.png 
 OK: My Local Test 0_4_bodycol-md-4nth-of-type3_1_tablet_v.png 
 ERROR: My Local Test 0_1_jumbotron_0_phone.png 
   See: /Users/sohail/GitHub/test/backstop_data/bitmaps_test/20160805-172309/failed_diff_0_1_jumbotron_0_phone.png
 ERROR: My Local Test 0_1_jumbotron_1_tablet_v.png 
   See: /Users/sohail/GitHub/test/backstop_data/bitmaps_test/20160805-172309/failed_diff_0_1_jumbotron_1_tablet_v.png
 ERROR: My Local Test 0_1_jumbotron_2_tablet_h.png 
   See: /Users/sohail/GitHub/test/backstop_data/bitmaps_test/20160805-172309/failed_diff_0_1_jumbotron_2_tablet_h.png

Test completed...
 15 Passed 
 3 Failed
 
 *** Mismatch errors found *** 
For a detailed report run `npm run openReport`

[17:23:16] 'compare' error after 522 ms
[17:23:16] Error: Mismatch errors found.
    at updateProgress (/Users/sohail/GitHub/test/node_modules/backstopjs/gulp/tasks/compare.js:37:22)
    at Array.<anonymous> (/Users/sohail/GitHub/test/node_modules/backstopjs/gulp/tasks/compare.js:77:13)
    at triggerDataUpdate (/Users/sohail/GitHub/test/node_modules/node-resemble-js/resemble.js:69:28)
    at onceWeHaveBoth (/Users/sohail/GitHub/test/node_modules/node-resemble-js/resemble.js:447:6)
    at null.<anonymous> (/Users/sohail/GitHub/test/node_modules/node-resemble-js/resemble.js:125:13)
    at emitOne (events.js:77:13)
    at emit (events.js:169:7)
    at null.<anonymous> (/Users/sohail/GitHub/test/node_modules/pngjs/lib/png.js:52:14)
    at emitOne (events.js:77:13)
    at emit (events.js:169:7)
[17:23:16] Starting 'openReport'...

Testing with  ../../backstop_data/bitmaps_test/compare.json
Opening report ->  http://localhost:3001/compare/

[17:23:16] Finished 'openReport' after 12 ms

npm ERR! Darwin 15.6.0
npm ERR! argh "/usr/local/bin/node" "/usr/local/bin/npm" "run" "test"
npm ERR! node v4.4.7
npm ERR! npm  v3.10.5
npm ERR! code ELIFECYCLE
npm ERR! backstopjs@1.3.5 test: `gulp test`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the backstopjs@1.3.5 test script 'gulp test'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the backstopjs package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     gulp test
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs backstopjs
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls backstopjs
npm ERR! There is likely additional logging output above.
npm WARN Local package.json exists, but node_modules missing, did you mean to install?

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/sohail/GitHub/test/node_modules/backstopjs/npm-debug.log
```

**Comment by nobso** [comment]
I also faced this issue before, @mig-25 Can you check those "failed_diff*" files to see if those failures are genuine?

**Comment by nobso** [comment]
It seems to be, the error reporting is intentional, I can see that the change was made by https://github.com/garris/BackstopJS/pull/215 

@garris @levithomason could you please let us know the rationale behind this change?

**Comment by garris** [comment]
@nobso Thanks for following up on this!   

Yes. It is intentional. But I regretting this change because of all the confusion it is creating 😞 .

It was changed because of CLI users who need the error to propagate all the way back to the original invocation so that the calling process can report a failure.  Some of these users were confused that they needed to add an extra step to enable error bubbling.

And of course, web developers are getting confused by all the npm messages.   At this point I prefer to add a new config `CLI_SILENT` and make that the default.   I think this may be more intuitive for regular users who will just use the default setting and users who will be integrating into a larger process will see the **_SILENT** and get a hint that there is a non-silent option.  

Or something like that...

Comments welcome!

**Comment by nobso** [comment]
@garris Thanks for the quick response.
How would this 'CLI_SILENT" be different from 'cliExitOnFail:true'?
IMHO, can't we just use the same "debug: true" flag to report the error stack back? in this way, we may not need to introduce an another config into the system.

**Comment by garris** [comment]
They all do the same thing.  The only difference is how users learn to use BackstopJS.   I thought originally that users would quickly see cliExitOnFail and think to switch it's state.  But there were multiple users who were didn't understand this.  My thinking is that someone setting up backstop for the first time is less likely to read the documentation and just want's it to work how they expect.  So,  my thinking is that eliminating the need to learn another parameter and putting the behavior into the name of the config flag would help.   ¯\_(ツ)_/¯

**Comment by nobso** [comment]
Yes, i agree with you.
But to do so, why can't we use the existing flag called 'debug'?

**Comment by garris** [comment]
Well, debug adds a lot more detail the CLI window.  It's purpose is to help figure out why BackstopJS isn't working.  So, it shouldn't change behavior for an CI implementation detail.

**Comment by nobso** [comment]
I see. Thanks for the clarification.

**Comment by levithomason** [comment]
### Why Error Codes?

>@garris @levithomason could you please let us know the rationale behind this change?

It is standard practice for  CLIs to return a non 0 code on failure.  It is considered a bug to return a 0 when the process encountered an error.  Exit codes are used by downstream processes to determine if some command completed successfully or not.  

The reason this came up mostly from users using Backstop in CI/CD systems, and not with other users, is because CI/CD systems rely on the CLIs returning proper error codes.  If exit code `0` (pass) is returned when tests failed, then the CI/CD service will ship your app with failing tests.

Any respectable CLI will exit with proper codes, like `git`:

```
in levithomason/  
› git status
fatal: Not a git repository (or any of the parent directories): .git

in levithomason/  
› echo $?
128
```

```
in stardust/ on feature/message-no-jquery 
› git branch -d feature/message-no-jquery
error: Cannot delete the branch 'feature/message-no-jquery' which you are currently on.

in stardust/ on feature/message-no-jquery 
› echo $?
1
```

NPM itself:

```
in levithomason/  
› npm run not-a-script --silent

in levithomason/  
› echo $?
254
```

This is also why the default NPM package.json test script exits 1:

```
  "test": "echo \"Error: no test specified\" && exit 1"
```

Gulp example:

```
in BackstopJS/
› $(npm bin)/gulp foo

[13:18:59] Using gulpfile ~/src/BackstopJS/gulpfile.js
[13:18:59] Task 'foo' is not in your gulpfile
[13:18:59] Please check the documentation for proper gulpfile formatting

in BackstopJS/
› echo $?
1
```

FWIW, all test frameworks and runners, like `mocha`, `ava`, `jasmine`, etc. will also exit with proper codes.  `0` on success, and something else for failures.

### What's with all the error output?

>But after the Finished 'openReport' after 12 ms lot of errors are reported as included in this message.

In short, there are 3 tools outputting errors.  Backstop, Gulp, NPM, in that order.  Once the CLI work is merged Gulp errors will no longer be a factor.

The lines starting with `npm ERR!` are standard NPM errors. They are unrelated to Backstop, gulp, and any other tool at play.  Since NPM is process being executed (`npm test`), NPM logs its own error on failure (exit non `0`) as well.

Any NPM command that fails gets this output.  If you'd like to hide those errors, you need to run `npm test --silent`, see [npm config](https://docs.npmjs.com/misc/config).  You can see more on NPM making this better in this issue https://github.com/npm/npm/issues/8821.  But again, this is completely unrelated to Backstop or any other tool in the chain.  It is just the way NPM works.

The stack trace error beginning with:

```
[17:23:16] Error: Mismatch errors found.
```

Can be removed since Backstop reports its own errors regarding the comparison failure.  This will likely be handled in the CLI work going on.  This error is reported twice because Backstop writes to the output, then gulp reports the error since the error is passed to the gulp callback:

```
compare.js
done(new Error('Mismatch errors found.'));
```

Once the CLI work is done, the whole experience will be fantastically improved.  There will proper exit codes, a top level `backstop` CLI (instead of nested gulp), and clean output reporting.

Hope I didn't come across and haughty here, but I really think the CLI work and dropping gulp is the key to polishing off this already great project.

**Comment by mig-25** [comment]
The test failures are genuine as they are supposed to after altering the css. But I am wondering, why is not launching the browser in my mac but spins upp the Windows virtual machine and starts chrome there but fails to load the page of course. Typing in the url on mac in safari doesn't load the page either. 
But if I manually launch file:///Users/sohail/GitHub/test/node_modules/backstopjs/compare/index.html
the report shows up with the proper tests failing.

**Comment by levithomason** [comment]
If the page is not served at `http://localhost:3001/compare/`, then it is likely that the test server is not running.  This might make sense if the gulp process stops due the error before it can start the server for you.

Did you try [running the report server](https://github.com/garris/BackstopJS#running-the-report-server) before opening the report?

```
$ npm start
$ npm run openReport
```

This will start the report server, then open the report page.

**Comment by garris** [comment]
Closed by #259
